tower-http/src/normalize_path.rs

```diff
@@ -0,0 +1,192 @@
+//!     .service_fn(handle);
+//!
+//! // call the service
+//! let request = Request::builder()
+//!     // `handle` will see `/foo`
+//!     .uri("/foo/")
+//!     .body(Body::empty())?;
+//!
+//! service.ready().await?.call(request).await?;
+//! #
+//! # Ok(())
+//! # }
+//! ```
+
+use http::{Request, Response, Uri};
+use std::{
+    borrow::Cow,
+    task::{Context, Poll},
+};
+use tower_layer::Layer;
+use tower_service::Service;
+
+/// Layer that applies [`NormalizePath`] which normalizes paths.
+///
+/// See the [module docs](self) for more details.
+#[derive(Debug, Copy, Clone)]
+pub struct NormalizePathLayer {}
+
+impl NormalizePathLayer {
+    /// Create a new [`NormalizePathLayer`].
+    ///
+    /// Any trailing slashes from request paths will be removed. For example, a request with `/foo/`
+    /// will be changed to `/foo` before reaching the inner service.
+    pub fn trim_trailing_slash() -> Self {
+        NormalizePathLayer {}
+    }
+}
+
+impl<S> Layer<S> for NormalizePathLayer {
+    type Service = NormalizePath<S>;
+
+    fn layer(&self, inner: S) -> Self::Service {
+        NormalizePath::trim_trailing_slash(inner)
+    }
+}
+
+/// Middleware that normalizes paths.
+///
+/// See the [module docs](self) for more details.
+#[derive(Debug, Copy, Clone)]
+pub struct NormalizePath<S> {
+    inner: S,
+}
+
+impl<S> NormalizePath<S> {
+    /// Create a new [`NormalizePath`].
+    ///
+    /// Any trailing slashes from request paths will be removed. For example, a request with `/foo/`
+    /// will be changed to `/foo` before reaching the inner service.
+    pub fn trim_trailing_slash(inner: S) -> Self {
+        Self { inner }
+    }
+
+    define_inner_service_accessors!();
+}
+
+impl<S, ReqBody, ResBody> Service<Request<ReqBody>> for NormalizePath<S>
+where
+    S: Service<Request<ReqBody>, Response = Response<ResBody>>,
+{
+    type Response = S::Response;
+    type Error = S::Error;
+    type Future = S::Future;
+
+    #[inline]
+    fn poll_ready(&mut self, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        self.inner.poll_ready(cx)
+    }
+
+    fn call(&mut self, mut req: Request<ReqBody>) -> Self::Future {
+        remove_trailing_slash(req.uri_mut());
+        self.inner.call(req)
+    }
+}
+
+fn remove_trailing_slash(uri: &mut Uri) {
+    if !uri.path().ends_with('/') {
+        return;
+    }
+
+    let new_path = uri.path().trim_end_matches('/');
+
+    let mut parts = uri.clone().into_parts();
+
+    let new_path_and_query = if let Some(path_and_query) = &parts.path_and_query {
+        let new_path_and_query = if let Some(query) = path_and_query.query() {
+            Cow::Owned(format!("{}?{}", new_path, query))
+        } else {
+            new_path.into()
+        }
+        .parse()
+        .unwrap();
+
+        Some(new_path_and_query)
+    } else {
+        None
+    };
+
+    parts.path_and_query = new_path_and_query;
+    *uri = Uri::from_parts(parts).unwrap();
```

Comment:
This is also vulnerable to https://github.com/tokio-rs/axum/issues/1122. I'll fix that before merging.

Reply:
I'm still not really clear on how that could happen. How can the server receive URIs without a scheme?

Reply:
Yeah that also confuses me but it happens https://discord.com/channels/500028886025895936/990823637084934144/991007953010188308. But in http maybe 🤷